src/background/host.ts

```diff
@@ -1,26 +1,87 @@
+import Debug from "debug";
 import Peer from "peerjs";
-import { Message, MessageType } from "../Message";
+import { Friend, LocalOutMessage, LocalPollMessage, MessageType, RemoteMessage } from "../Message";
+
+export interface FriendConnected extends Friend {
+    conn: Peer.DataConnection;
+}
 
 export class Host {
+    #log = Debug("peer:disconnected");
     #port: chrome.runtime.Port;
     #peer = new Peer();
-    #friends = new Set<Peer.DataConnection>();
 
-    public isReady = false;
-    public id: string | null = null;
+    #friends = new Map<string, FriendConnected>();
+
+    public personalData: Friend | null = null;
     // eslint-disable-next-line @typescript-eslint/no-empty-function
     public onChangeFriends: (count: number, delta: number) => void = () => {};
 
-    #handleConnect = (conn: Peer.DataConnection): void => {
-        this.#friends.add(conn);
-        this.onChangeFriends(this.#friends.size, 1);
+    #ensureConnections = (friends: Friend[], conn: Peer.DataConnection): void => {
+        if (!this.personalData) {
+            return;
+        }
+
+        for (const member of friends) {
+            const friend = this.#friends.get(member.id);
+
+            if (member.id === this.personalData.id) {
+                // do nothing regarding our own connection
+            } else if (friend) {
+                // update member data without touching `friend.conn`
+                this.#friends.set(member.id, {
+                    ...member,
+                    ...friend
+                });
+            } else if (member.id === conn.peer) {
+                // insert the person we connected to, now that we know their info
+                this.#friends.set(conn.peer, {
+                    ...member,
+                    conn
+                });
+                this.onChangeFriends(this.#friends.size, 1);
+            } else {
+                // for new people, just go connect to them (& hopefully they aren't doing the same)
+                this.connect(member.id);
+            }
+        }
+    };
+
+    #handleConnect = (conn: Peer.DataConnection, isIncoming: boolean): void => {
+        if (!this.personalData) {
+            return;
+        }
+
+        this.#log(`Handling ${isIncoming ? "in" : "out"} connection`, conn.metadata);
+
+        if (isIncoming) {
+            // `conn.metadata` will always be the `Friend` of the **sender**, so when we handle the outgoing request,
+            // we avoid writing ourselves as a friend (and therefore emitting events to ourselves & getting
+            // infinite-looped).
+            const friend: FriendConnected = {
+                ...(conn.metadata as Friend),
+                conn
+            };
+
+            this.#friends.set(conn.peer, friend);
+            this.onChangeFriends(this.#friends.size, 1);
```

Comment:
This is where I got stuck and introduced the `Math.random()`, which isn't perfect. The random works fine as long as they don't pick the same `~40ms` window.

I'm not really even sure what the solution would be ... since we wouldn't know which connection to destroy and which one to keep. B/c if the other side doesn't make the same choice, you've then destroyed both connections, lol.

Reply:
Got this fixed using a deterministic comparison (`id` alphanumerically)